### PR TITLE
Connect remote clients

### PR DESCRIPTION
**Pull request by justinmoon**
This PR adds the ability to connect to a federation and download it's config.

To do this I create a struct called `WsFederationConnect`. This just contains enough information to instantiate `WsFederationApi`.

I also try to convert some `String`s to `SocketAddr`s in pursuit of https://github.com/fedimint/fedimint/issues/376.

### CODE REVIEW

File: client/client-lib/src/api.rs [hunk 1]
```diff
@@ -158,6 +160,30 @@ struct FederationMember<C> {
...
 }
 
+/// Information required for client to construct [`WsFederationApi`] instance
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WsFederationConnect {
+    pub members: Vec<(PeerId, String)>,
+    pub max_evil: usize,
+}
+
+impl From<ClientConfig> for WsFederationConnect {
```
Comment (elsirion) [inline client/client-lib/src/api.rs:170]: Shouldn't that be `From<&ClientConfig>`?

**Review by elsirion — APPROVED** [review]
Left some nits, up to you if we want to merge and fix later or just do some quick fixes.

File: minimint/src/net/api.rs [hunk 2]
```diff
@@ -134,6 +139,32 @@ fn server_endpoints() -> &'static [ApiEndpoint<MinimintConsensus<rand::rngs::OsR
...
+                    peg_in_descriptor: minimint.cfg.wallet.peg_in_descriptor.clone(),
+                    network: minimint.cfg.wallet.network,
+                    fee_consensus: minimint.cfg.wallet.fee_consensus.clone(),
+                    finality_delay: minimint.cfg.wallet.finality_delay,
+                };
+                let ln = LightningModuleClientConfig {
+                    threshold_pub_key: minimint.cfg.ln.threshold_pub_keys.public_key(),
+                    fee_consensus: minimint.cfg.ln.fee_consensus.clone(),
+                };
+                Ok(ClientConfig { api_endpoints, mint, wallet, ln, max_evil }) }
```
[the remark this answers is not in the document]
Comment (justinmoon) [reply minimint/src/net/api.rs]: Done

File: minimint/src/net/api.rs [hunk 3]
```diff
@@ -134,6 +139,32 @@ fn server_endpoints() -> &'static [ApiEndpoint<MinimintConsensus<rand::rngs::OsR
...
+            async |minimint: &MinimintConsensus<rand::rngs::OsRng>, _v: ()| -> ClientConfig {
+                let api_endpoints: Vec<String> = minimint
+                    .cfg
+                    .peers
+                    .iter()
+                    .map(|(_, peer)| peer.connection.api_addr.clone())
+                    .collect();
+                let max_evil = hbbft::util::max_faulty(minimint.cfg.peers.len());
+                let mint = MintClientConfig {
+                    tbs_pks: Keys::from_iter(minimint.mint.pub_key().into_iter()),
```
[the remark this answers is not in the document]
Comment (justinmoon) [reply minimint/src/net/api.rs]: Made an issue https://github.com/fedimint/fedimint/issues/402

File: ln-gateway/src/bin/ln_gateway.rs [hunk 4]
```diff
@@ -71,6 +72,17 @@ async fn initialize_gateway(
...
         _ => unreachable!(),
     };
+    let host = match plugin.option("minimint-host") {
+        Some(options::Value::String(workdir)) => workdir,
+        _ => unreachable!(),
+    };
+    let port = match plugin.option("minimint-port") {
+        Some(options::Value::String(workdir)) => workdir,
+        _ => unreachable!(),
+    };
```
Comment (justinmoon) [inline ln-gateway/src/bin/ln_gateway.rs:82]: This is so ugly. Is the a more idiomatic way to do this?

**Comment by justinmoon** [comment]
Should be ready to go. Sorry this took so long!

File: modules/minimint-mint/src/config.rs [hunk 5]
```diff
@@ -73,6 +77,25 @@ impl GenerateConfig for MintConfig {
...
+
+    fn to_client_config(&self) -> Self::ClientConfig {
+        let pub_key: HashMap<Amount, AggregatePublicKey> = TieredMultiZip::new(
+            self.peer_tbs_pks
+                .iter()
+                .map(|(_, keys)| keys.iter())
+                .collect(),
+        )
+        .map(|(amt, keys)| {
+            // TODO: avoid this through better aggregation API allowing references or
```
Comment (elsirion) [inline modules/minimint-mint/src/config.rs:89]: Was there something to follow after the "or"?

**Review by elsirion — APPROVED** [review]

File: modules/minimint-mint/src/config.rs (hunk 5 above)
Comment (elsirion) [reply modules/minimint-mint/src/config.rs:89]: Just a nit, merging anyway

Comment (justinmoon) [reply modules/minimint-mint/src/config.rs:89]: Ask the guy who wrote it ;) https://github.com/fedimint/fedimint/blob/f91fb2fa30e4e8e0e1a9ac8bc6f410ed07171bcd/modules/minimint-mint/src/lib.rs#L437

**Pull request merged**